include/gridtools/fn/backend2/gpu.hpp

```diff
@@ -0,0 +1,165 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include "../../common/cuda_util.hpp"
+#include "../../common/hymap.hpp"
+#include "../../meta.hpp"
+#include "../../sid/allocator.hpp"
+#include "../../sid/concept.hpp"
+#include "../../sid/contiguous.hpp"
+#include "../../sid/multi_shift.hpp"
+#include "./common.hpp"
+
+namespace gridtools::fn::backend {
+    namespace gpu_impl_ {
+        template <class BlockSizes>
+        struct gpu {
+            using block_sizes_t = BlockSizes;
+        };
+
+        template <class BlockSizes, class Dims, int I>
+        using block_size_at_dim = meta::second<meta::mp_find<BlockSizes, meta::at_c<Dims, I>>>;
+
+        template <class BlockSizes, class Sizes>
+        GT_FUNCTION_DEVICE auto global_thread_index() {
+            using ndims_t = meta::length<Sizes>;
+            using keys_t = meta::take_c<std::min(3, (int)ndims_t::value), get_keys<Sizes>>;
+            if constexpr (ndims_t::value == 0) {
+                return hymap::keys<>::values<>();
+            } else if constexpr (ndims_t::value == 1) {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using values_t = typename keys_t::template values<int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x);
+            } else if constexpr (ndims_t::value == 2) {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
+                using values_t = typename keys_t::template values<int, int>;
+                return values_t(
+                    blockIdx.x * block_dim_x::value + threadIdx.x, blockIdx.y * block_dim_y::value + threadIdx.y);
+            } else {
+                using block_dim_x = block_size_at_dim<BlockSizes, keys_t, 0>;
+                using block_dim_y = block_size_at_dim<BlockSizes, keys_t, 1>;
+                using block_dim_z = block_size_at_dim<BlockSizes, keys_t, 2>;
+                using values_t = typename keys_t::template values<int, int, int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x,
+                    blockIdx.y * block_dim_y::value + threadIdx.y,
+                    blockIdx.z * block_dim_z::value + threadIdx.z);
+            }
+        }
+
+        template <class Index, class Sizes>
+        GT_FUNCTION_DEVICE bool in_domain(Index const &index, Sizes const &sizes) {
+            return tuple_util::device::all_of(std::less(), index, sizes);
```

Comment:
now `index` dimensions is sripped to 3 and the `sizes` dimensions are not. It will cause compile time error in multi dim case.

Reply:
It actually works as the implementation of `all_of` takes the iteration range from the first tuple’s size. But probably that’s an implementation detail I should not rely on…